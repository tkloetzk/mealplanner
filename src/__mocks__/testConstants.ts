// src/test/testConstants.ts
import { CategoryType, ServingSizeUnit } from "@/types/food";
import { Kid } from "@/types/user";
import { BREAKFAST, LUNCH, DINNER, DAYS_OF_WEEK } from "@/constants";
import { RANCH_OPTION } from "@/constants/meal-goals";
// Constant for selected day
export const SELECTED_DAY = DAYS_OF_WEEK[6];

// Mock Kids
export const MOCK_KIDS: Kid[] = [
  { id: "kid1", name: "Child One" },
  { id: "kid2", name: "Child Two" },
];

// Category Types
export const PROTEINS = "proteins";
export const FRUITS = "fruits";
export const VEGETABLES = "vegetables";
export const GRAINS = "grains";
export const OTHER = "other";

// Mock Foods
export const MOCK_FOODS = {
  proteins: [
    {
      id: "1",
      name: "Chicken Breast",
      category: PROTEINS,
      calories: 165,
      protein: 31,
      carbs: 0.1,
      fat: 3.6,
      meal: [BREAKFAST, LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "piece" as ServingSizeUnit,
      hiddenFromKids: false,
    },
    {
      id: "4",
      name: "Not a breakfast food",
      category: "proteins" as CategoryType,
      calories: 100,
      protein: 30,
      carbs: 0.1,
      fat: 34.6,
      meal: [LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "piece" as ServingSizeUnit,
    },
    // Add more protein foods
  ],
  fruits: [
    {
      id: "2",
      name: "Apple",
      category: "fruits" as const,
      calories: 95,
      protein: 0.5,
      carbs: 25,
      fat: 0.3,
      meal: [BREAKFAST, LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "piece" as ServingSizeUnit, // Explicitly type as ServingSizeUnit
    },
    // Add more fruit foods
  ],
  vegetables: [
    {
      id: "3",
      name: "Broccoli",
      category: "vegetables" as CategoryType,
      calories: 55,
      protein: 4,
      carbs: 11,
      fat: 0.6,
      meal: [LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "cup" as ServingSizeUnit,
    },
    // Add more vegetable foods
  ],
  other: [
    {
      id: "5",
      name: "Mixed Snack Plate",
      calories: 200,
      protein: 5,
      carbs: 25,
      fat: 10,
      servingSize: "1",
      servingSizeUnit: "plate" as ServingSizeUnit,
      category: "other" as CategoryType,
      meal: [BREAKFAST, LUNCH, DINNER, "snack"],
      hiddenFromChild: true,
    },
    {
      id: "6",
      name: "Visible Snack",
      calories: 150,
      protein: 3,
      carbs: 20,
      fat: 5,
      servingSize: "1",
      servingSizeUnit: "serving" as ServingSizeUnit,
      category: "other" as CategoryType,
      meal: [BREAKFAST, LUNCH],
      hiddenFromChild: false,
    },
  ],
  condiments: [RANCH_OPTION],
};
