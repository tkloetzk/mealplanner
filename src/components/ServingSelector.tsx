// components/ServingSelector.tsx
import React, { useState } from "react";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Minus, Plus, Check, X } from "lucide-react";
import { Food, SelectedFood } from "@/types/food";

const PRESET_SERVINGS = [
  { label: "¼", value: 0.25 },
  { label: "½", value: 0.5 },
  { label: "¾", value: 0.75 },
  { label: "1", value: 1 },
  { label: "1½", value: 1.5 },
  { label: "2", value: 2 },
];

// In ServingSelector.tsx
interface ServingSelectorProps {
  food: Food;
  currentServings?: number; // Add this prop
  onConfirm: (adjustedFood: SelectedFood) => void;
  onCancel: () => void;
}

export function ServingSelector({
  food,
  currentServings = 1,
  onConfirm,
  onCancel,
}: ServingSelectorProps) {
  // Initialize with currentServings if provided, otherwise use 1
  const [servings, setServings] = useState(currentServings);

  const adjustNutrition = (servingCount: number): SelectedFood => {
    return {
      ...food,
      servings: servingCount,
      adjustedCalories: food.calories * servingCount,
      adjustedProtein: food.protein * servingCount,
      adjustedCarbs: food.carbs * servingCount,
      adjustedFat: food.fat * servingCount,
    };
  };

  const handleServingChange = (newServings: number) => {
    if (newServings >= 0.25 && newServings <= 5) {
      setServings(newServings);
    }
  };

  return (
    <div className="fixed inset-0 bg-black/50 flex items-center justify-center p-4 z-50">
      <div className="bg-white rounded-lg w-full max-w-md p-4 md:p-6">
        <div className="flex justify-between items-center mb-4">
          <h3 className="text-lg font-semibold">{food.name}</h3>
          <Button variant="ghost" size="icon" onClick={onCancel}>
            <X className="h-4 w-4" />
          </Button>
        </div>

        <div className="text-sm text-gray-600 mb-4">
          Serving size: {food.servingSize} {food.servingSizeUnit}
        </div>

        {/* Preset Servings */}
        <div className="grid grid-cols-3 sm:grid-cols-6 gap-2 mb-4">
          {PRESET_SERVINGS.map(({ label, value }) => (
            <Button
              key={value}
              variant={servings === value ? "default" : "outline"}
              className="w-full"
              onClick={() => handleServingChange(value)}
            >
              {label}
            </Button>
          ))}
        </div>

        {/* Custom Serving Input */}
        <div className="flex items-center gap-2 mb-6">
          <Button
            variant="outline"
            size="icon"
            onClick={() => handleServingChange(servings - 0.25)}
          >
            <Minus className="h-4 w-4" />
          </Button>

          <Input
            type="number"
            value={servings}
            step="0.25"
            min="0.25"
            max="5"
            className="w-20 text-center"
            data-testid="custom-serving-input"
            onChange={(e) => handleServingChange(parseFloat(e.target.value))}
          />

          <Button
            variant="outline"
            size="icon"
            data-testid="increment-serving"
            onClick={() => handleServingChange(servings + 0.25)}
          >
            <Plus className="h-4 w-4" />
          </Button>
        </div>

        {/* Nutrition Info */}
        <div className="grid grid-cols-2 gap-4 mb-6 p-4 bg-gray-50 rounded-lg text-sm">
          <div className="flex justify-between">
            <span>Calories:</span>
            <span className="font-medium" data-testid="calories-incremented">
              {Math.round(food.calories * servings)}
            </span>
          </div>
          <div className="flex justify-between">
            <span>Protein:</span>
            <span className="font-medium">
              {(food.protein * servings).toFixed(1)}g
            </span>
          </div>
          <div className="flex justify-between">
            <span>Carbs:</span>
            <span className="font-medium">
              {(food.carbs * servings).toFixed(1)}g
            </span>
          </div>
          <div className="flex justify-between">
            <span>Fat:</span>
            <span className="font-medium">
              {(food.fat * servings).toFixed(1)}g
            </span>
          </div>
        </div>

        <div className="flex justify-end gap-2">
          <Button variant="outline" onClick={onCancel}>
            Cancel
          </Button>
          <Button onClick={() => onConfirm(adjustNutrition(servings))}>
            <Check className="h-4 w-4 mr-2" />
            Confirm
          </Button>
        </div>
      </div>
    </div>
  );
}
