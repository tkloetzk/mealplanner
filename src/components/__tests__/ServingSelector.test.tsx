// src/components/__tests__/ServingSelector.test.tsx
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { ServingSelector } from "../ServingSelector";
import { MOCK_FOODS } from "@/constants/tests/testConstants";

describe("ServingSelector Component", () => {
  const mockOnConfirm = jest.fn();
  const mockOnCancel = jest.fn();

  const defaultProps = {
    food: MOCK_FOODS.proteins[0],
    currentServings: 1,
    onConfirm: mockOnConfirm,
    onCancel: mockOnCancel,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders initial serving correctly", () => {
    render(<ServingSelector {...defaultProps} />);

    // Check initial serving input
    const servingInput = screen.getByDisplayValue("1");
    expect(servingInput).toBeInTheDocument();
  });

  it("calculates nutrition for different servings", () => {
    render(<ServingSelector {...defaultProps} />);

    const caloriesElement = screen.getByTestId("calories-incremented");
    expect(caloriesElement).toBeInTheDocument();
    expect(caloriesElement).toHaveTextContent(
      MOCK_FOODS.proteins[0].calories.toString()
    );
    // Find increment button and click
    const incrementButton = screen.getByTestId("increment-serving");
    fireEvent.click(incrementButton);

    // Check updated nutrition values
    expect(caloriesElement).toHaveTextContent(
      (MOCK_FOODS.proteins[0].calories * 1.25).toFixed(0).toString()
    );
  });

  it("handles confirmation", () => {
    render(<ServingSelector {...defaultProps} />);

    const confirmButton = screen.getByRole("button", { name: /confirm/i });
    fireEvent.click(confirmButton);

    expect(mockOnConfirm).toHaveBeenCalledTimes(1);
  });

  it("handles cancellation", () => {
    render(<ServingSelector {...defaultProps} />);

    const cancelButton = screen.getByRole("button", { name: /cancel/i });
    fireEvent.click(cancelButton);

    expect(mockOnCancel).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<ServingSelector {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
