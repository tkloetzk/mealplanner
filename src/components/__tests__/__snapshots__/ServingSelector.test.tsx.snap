// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServingSelector Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-0 bg-black/50 flex items-center justify-center p-4 z-50"
  >
    <div
      class="bg-white rounded-lg w-full max-w-md p-4 md:p-6"
    >
      <div
        class="flex justify-between items-center mb-4"
      >
        <h3
          class="text-lg font-semibold"
        >
          Chicken Breast
        </h3>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-9 w-9"
        />
      </div>
      <div
        class="text-sm text-gray-600 mb-4"
      >
        Serving size: 1 piece
      </div>
      <div
        class="grid grid-cols-3 sm:grid-cols-6 gap-2 mb-4"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-full"
        >
          ¼
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-full"
        >
          ½
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-full"
        >
          ¾
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2 w-full"
        >
          1
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-full"
        >
          1½
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-full"
        >
          2
        </button>
      </div>
      <div
        class="flex items-center gap-2 mb-6"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 w-9"
        />
        <input
          class="flex h-9 rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm w-20 text-center"
          max="5"
          min="0.25"
          step="0.25"
          type="number"
          value="1"
        />
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 w-9"
          data-testid="increment-serving"
        />
      </div>
      <div
        class="grid grid-cols-2 gap-4 mb-6 p-4 bg-gray-50 rounded-lg text-sm"
      >
        <div
          class="flex justify-between"
        >
          <span>
            Calories:
          </span>
          <span
            class="font-medium"
            data-testid="calories-incremented"
          >
            165
          </span>
        </div>
        <div
          class="flex justify-between"
        >
          <span>
            Protein:
          </span>
          <span
            class="font-medium"
          >
            31.0g
          </span>
        </div>
        <div
          class="flex justify-between"
        >
          <span>
            Carbs:
          </span>
          <span
            class="font-medium"
          >
            0.0g
          </span>
        </div>
        <div
          class="flex justify-between"
        >
          <span>
            Fat:
          </span>
          <span
            class="font-medium"
          >
            3.6g
          </span>
        </div>
      </div>
      <div
        class="flex justify-end gap-2"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
        >
          Cancel
        </button>
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
        >
          Confirm
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
